experimental/redux/otelMiddleware.ts

```diff
@@ -0,0 +1,42 @@
+import {
+  Dispatch,
+  Middleware,
+  MiddlewareAPI,
+  Action,
+  UnknownAction,
+} from "redux";
+import {TracerProvider, trace} from "@opentelemetry/api";
+
+interface ConfigArgs {
+  name: string;
+  version: string;
+}
+
+const otelMiddleware = <RootState>(
+  provider: TracerProvider,
+  config: ConfigArgs,
+): Middleware<object, RootState> => {
+  if (!provider) {
+    console.warn("Provider is not available. Spans won't be recorded");
+  }
+
+  const tracer = trace.getTracer(config.name, config.version);
+
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-expect-error
+  return (_: MiddlewareAPI<Dispatch<UnknownAction>, RootState>) => {
+    return (next: Dispatch<UnknownAction>) => {
+      return (action: Action) => {
+        const span = tracer.startSpan("redux-action");
```

Comment:
taking a look at how middleware works we should maybe revisit the instrumentation here, it's a bit unfortunate that it works as a chain so it seems you really only get insight into how long the "next" middleware took but not how long the whole action took to work through the dispatch operation

there's maybe some potential alternatives with either a store enhancer or subscribing to change events:
https://redux.js.org/understanding/thinking-in-redux/glossary#store-enhancer
https://redux.js.org/api/store#subscribelistener
https://github.com/reduxjs/redux/issues/303#issuecomment-125184409

or if we kept to this approach maybe makes more sense as a log instead of span to record the single event